lib/api/units/validateBody.js

```diff
@@ -0,0 +1,233 @@
+const jph = require('json-parse-helpfulerror');
+const mediaTyper = require('media-typer');
+const contentTypeUtils = require('content-type');
+
+const { TextDiff } = require('../../../lib/validators/text-diff');
+const { JsonExample } = require('../../../lib/validators/json-example');
+const { JsonSchema } = require('../../../lib/validators/json-schema');
+
+function isPlainText(mediaType) {
+  return mediaType.type === 'text' && mediaType.subtype === 'plain';
+}
+
+function isJson(mediaType) {
+  return mediaType.type === 'application' && mediaType.subtype === 'json';
+}
+
+function isJsonWeak(mediaType) {
+  return isJson(mediaType) || mediaType.suffix === 'json';
+}
+
+/**
+ * Parses a given content-type header into media type.
+ * @param {string} contentType
+ * @returns {[Error, MediaType]}
+ */
+function parseContentType(contentType) {
+  try {
+    const { type } = contentTypeUtils.parse(`${contentType}`);
+    const mediaType = mediaTyper.parse(type);
+    return [null, mediaType];
+  } catch (error) {
+    return [error, null];
+  }
+}
```

Comment:
My main concern is how broad `try/catch` can be applied as opposed to tuples.

I strongly dislike nesting of try/catch blocks, as it makes code unreadable:

https://github.com/apiaryio/gavel.js/blob/f58628f4cd2d837d6033f4abad99a3cd60d26b19/lib/mixins/validatable-http-message.js#L238-L289

Moreover, it makes error context matter, as if we throw in a wrong closure, a wrong `catch` handler will process it. Add it to combinable nature of try/catch blocks (children exception will be handled by a parent's `catch` block, if no closer `catch` block is met), and we also get high chances of bugs, with errors of one kind processed as if they are errors of the other.

Tuples, although verbose, do not introduce a separate closure for errors, and cannot be messed up with nesting. Also, my notion to have context-less errors is to change the validation output of Gavel in the future. I don't think it's correct to mix library-level warnings/errors (like "can't validate") with the actual validation errors (the diff). We can use a dedicated output collector for that, and there we would use errors in different context.

Reply:
> I strongly dislike nesting of try/catch blocks

I agree, we shouldn't do that.

> children exception will be handled by a parent's `catch` block

That's an unfortunate feature of how JavaScript works (you cannot filter by exception type in `catch`), yes, but I don't think it means we should throw the whole thing away because of it. It just means we need to be cautious when using the exceptions. Also generally, this is rather a feature than a bug. If you don't handle the exception, it crashes the software, which is good - that's fail-fast. So problem with parent blocks catching more than they should happens only if you're silencing errors. And one should be always very careful when doing that. Either it's expected, well contained behavior (if that errors, this defaults to XYZ), or it is top-level catch-all thing, which serializes the error to something user can read.

And there's always bug potential with error handling. E.g. if you forget to handle an error from a tuple, or a callback argument.

I'd say a good rule of thumb, at least in the code in this file, would be that we have small, dedicated helper functions, which throw errors. No nesting etc. The code using them should always be cautious to handle the error and decide what to do with it. In the example above I wrote it is well-contained, expected, even documented behavior, and there's a default value returned if the exception happens. IMHO if we keep using exceptions like that, it is a powerful tool.

> I don't think it's correct to mix library-level warnings/errors (like "can't validate") with the actual validation errors (the diff)

Agreed. But then it seems to me that all the helpers should use library-level errors and `validateBody()` has the responsibility to assess them and decide what does it mean for the user - either providing defaults, verdicts, or creating user-facing errors for the `results` array.

Reply:
How would you recommend to handle `getBodyType` with try/catch?

When I'm doing:

```js
try {
  jph.parse(...) // attempt to parse, throw if not parsable, OK
  const mediaType = parseContentType(...) // this may throw too
} catch(error) {
 // handle non-parsable json
}
```

If `parseContentType` call throws, the logic would treat current body as non-parsable JSON, since the error will be inherited and processed by the underlying catch block.

Reply:
Okay, I've reworked it under https://github.com/apiaryio/gavel.js/pull/155/commits/9c50d2f1216b5e1d47ed7d1eaaab28bf8db00fca.
I can see that it bypassed `null` values even previously, so in case of an error `parseContentType` can just return `null`. Seems like a reasonable compromise to me, since invalid content type is not an issue, it gets fallback to `application/json` exactly for that purpose.

In that block preceding `isJsonContentType` also calls `parseContentType`, so in case it fails, it would return `false` and the `application/json` fallback value would be used.

Reply:
My take would be something along the lines of:

```js
// Because "is this body parseable as JSON" is the main piece of information
// we're trying to get and convey in the mentioned code. Easy to unittest this way.
function isParseableJSON(body) {
  try {
    jph.parse(body);
    return true;
  } catch (error) {
    return false;
  }
}

...

// Implementation of the table (roughly) I drafted in the other comment, unittestable.
function getBodyType(contentType, bodyIsJSON) {
  if (!bodyIsJSON) return { type: 'text', subtype: 'plain' }; // effectively, if the content cannot be parsed as JSON, we're going to validate it as if it was plain text
  try {
    const mediaType = parseContentType(contentType);
    if (isJSON(mediaType)) return mediaType; // keeping application/hal+json and friends ... 🤔 is this actually useful for anything? the "effective" media type for validation is application/json
    return { type: 'application', subtype: 'json' }; // forcing the validation to process the body as if it was JSON
  } catch (error) {
    return bodyIsJSON
      ? { type: 'application', subtype: 'json' }
      : { type: 'text', subtype: 'plain' };
  }
}
```

Could be used as:

```js
const bodyType = getBodyType(headers['content-type'], isParseableJSON(body));
```

This is not a complete code, but you probably get the idea. Split to more contained unittestable functions, which do look nice even with exceptions in place. The surrounding code then only composes them together with occasional if and else. The reader doesn't have to understand how `isParseableJSON()` works, the name is enough.

One thing which occurred to me when assembling the code above is whether it actually helps with anything to treat the `+json` media types as special snowflakes. Or why is the decision so complicated? Perhaps the rule is as simple as "If Gavel can parse the body as JSON, it validates it as JSON. Otherwise it validates it as plain text."

I might be wrong, I didn't read all the code, but perhaps we don't have to fiddle with the content type at all if it doesn't effectively change anything?

```js
function validateBody(...) {
  ...

  if (isParseableJSON(body) {
    return ... // parse as if it was application/json
  } else {
    return ... // parse as if it was text/plain
  }
}
```